### ISSUE TITLE
High Priority: Replace Hardcoded Memory Requirements with Dynamic Profiling

### ISSUE THREAD

**Issue by jbmiller10** [jbmiller10/semantik#167]
## Summary
The system uses static, hardcoded memory requirements for models that don't reflect actual usage patterns. This leads to incorrect memory estimates, failed operations, and poor resource utilization. Dynamic profiling based on real measurements is needed.

## Severity
**High** - Causes incorrect resource allocation and operational failures

## Current Behavior
In `packages/vecpipe/memory_utils.py:12-33`, memory requirements are hardcoded:

```python
MODEL_MEMORY_REQUIREMENTS = {
    ("Qwen/Qwen3-Embedding-0.6B", "float32"): 2400,
    ("Qwen/Qwen3-Embedding-0.6B", "float16"): 1200,
    ("Qwen/Qwen3-Embedding-0.6B", "int8"): 600,
    # ... more hardcoded values
}
```

Problems with this approach:
- Values don't match actual memory usage
- No accounting for PyTorch overhead
- Ignores batch size impact
- Doesn't consider sequence length variations
- No adaptation to hardware differences

## Expected Behavior
The system should:
1. Profile actual memory usage on first load
2. Cache profiling results for future use
3. Account for dynamic factors (batch size, sequence length)
4. Update estimates based on runtime observations
5. Provide confidence intervals for estimates

## Impact
- **Failed Operations**: Underestimated memory causes OOM
- **Resource Waste**: Overestimated memory limits throughput
- **Poor Planning**: Can't accurately predict capacity
- **Inflexibility**: Can't adapt to new models or hardware

## Real-World Evidence
Based on production observations:
- Actual memory usage varies 30-50% from hardcoded values
- Batch size can change memory by 2-3x
- Sequence length has quadratic impact on attention memory
- PyTorch overhead adds 20-40% on top of model size

### 1. Dynamic Memory Profiler
```python
class ModelMemoryProfiler:
    """Profile and track actual model memory usage"""
    
    def __init__(self, cache_dir: str = "/tmp/model_memory_cache"):
        self.cache_dir = Path(cache_dir)
        self.cache_dir.mkdir(exist_ok=True)
        self.runtime_observations = deque(maxlen=1000)
        
    async def profile_model(self, model_name: str, quantization: str) -> ModelMemoryProfile:
        """Profile actual memory usage of a model"""
        
        # Check cache first
        cached = self._load_cached_profile(model_name, quantization)
        if cached and not cached.is_stale():
            return cached
        
        # Profile different scenarios
        profile = ModelMemoryProfile(model_name, quantization)
        
        # Test different sequence lengths and batch sizes
        test_lengths = [128, 512, 1024, 2048]
        test_batch_sizes = [1, 8, 16, 32]
        
        for seq_len in test_lengths:
            for batch_size in test_batch_sizes:
                try:
                    memory_mb = await self._measure_memory_usage(
                        model_name, quantization, batch_size, seq_len
                    )
                    
                    profile.add_measurement(
                        batch_size=batch_size,
                        sequence_length=seq_len,
                        memory_mb=memory_mb
                    )
                except Exception as e:
                    logger.warning(f"Failed to profile {batch_size}x{seq_len}: {e}")
        
        # Fit regression model for predictions
        profile.fit_model()
        
        # Save to cache
        self._save_profile(profile)
        
        return profile
```

### 2. Memory Usage Model
```python
class ModelMemoryProfile:
    """Model memory usage patterns"""
    
    def fit_model(self):
        """Fit regression model to predict memory usage"""
        
        # Prepare features
        X = []
        y = []
        
        for m in self.measurements:
            # Features: batch_size, seq_len, seq_len^2 (attention)
            X.append([
                m['batch_size'],
                m['sequence_length'],
                m['sequence_length'] ** 2 / 1000,  # Scaled quadratic term
                m['batch_size'] * m['sequence_length'] / 1000  # Interaction
            ])
            y.append(m['memory_mb'])
        
        # Fit linear regression
        from sklearn.linear_model import LinearRegression
        model = LinearRegression()
        model.fit(X, y)
        
        self.model_params = {
            'coefficients': model.coef_.tolist(),
            'intercept': model.intercept_,
            'r2_score': model.score(X, y)
        }
    
    def predict_memory(self, batch_size: int, sequence_length: int) -> tuple[float, float, float]:
        """Predict memory usage with confidence interval"""
        
        features = [
            batch_size,
            sequence_length,
            sequence_length ** 2 / 1000,
            batch_size * sequence_length / 1000
        ]
        
        predicted = (
            self.model_params['intercept'] +
            sum(c * f for c, f in zip(self.model_params['coefficients'], features))
        )
        
        # Add confidence interval based on R² score
        uncertainty = (1 - self.model_params['r2_score']) * 0.3
        lower = predicted * (1 - uncertainty)
        upper = predicted * (1 + uncertainty)
        
        return lower, predicted, upper
```

### 3. Runtime Memory Tracking
```python
class RuntimeMemoryTracker:
    """Track actual memory usage during operations"""
    
    def record_observation(self, model_name: str, quantization: str, 
                         batch_size: int, avg_sequence_length: int, 
                         actual_memory_mb: float):
        """Record actual memory usage"""
        
        key = (model_name, quantization)
        self.observations[key].append({
            'batch_size': batch_size,
            'sequence_length': avg_sequence_length,
            'memory_mb': actual_memory_mb,
            'timestamp': time.time()
        })
        
        # Update profile if we have enough new observations
        if len(self.observations[key]) >= 50:
            asyncio.create_task(self._update_profile(model_name, quantization))
```

### 4. Integration with Memory Checks
```python
def get_model_memory_requirement(model_name: str, quantization: str, 
                               batch_size: int = 1, 
                               max_sequence_length: int = 512) -> tuple[float, float, float]:
    """Get memory requirement with confidence bounds"""
    
    # Try to get profiled estimate
    if memory_profiler:
        try:
            profile = memory_profiler.get_cached_profile(model_name, quantization)
            
            if profile:
                return profile.predict_memory(batch_size, max_sequence_length)
        except Exception as e:
            logger.warning(f"Failed to get profiled memory: {e}")
    
    # Fallback to hardcoded with uncertainty
    base = MODEL_MEMORY_REQUIREMENTS.get((model_name, quantization), 2000)
    
    # Add factors for batch size and sequence length
    multiplier = 1 + (batch_size - 1) * 0.1 + (max_sequence_length / 512) * 0.2
    estimated = base * multiplier * MEMORY_OVERHEAD_FACTOR
    
    # Return with ±30% uncertainty
    return estimated * 0.7, estimated, estimated * 1.3
```

### Files to Modify
1. `packages/vecpipe/memory_utils.py` - Replace static dict with profiler
2. Create `packages/vecpipe/memory_profiler.py` - Profiling system
3. Update all memory estimation calls throughout codebase
4. Add profiling to model initialization

### Dependencies
- scikit-learn for regression modeling (already in requirements)
- Existing PyTorch memory tracking APIs

### Testing Requirements
1. **Accuracy Tests**: Compare estimates vs actual usage
2. **Profile Tests**: Verify profiling on different hardware
3. **Update Tests**: Test profile updates from runtime data
4. **Fallback Tests**: Behavior when profiling fails
5. **Performance Tests**: Profiling overhead

## Definition of Done
- [ ] Dynamic profiling system implemented
- [ ] Runtime observation and model updates
- [ ] Memory estimates within 20% of actual usage
- [ ] Confidence intervals on all estimates
- [ ] Monitoring of estimate accuracy
- [ ] Profile caching and persistence
- [ ] Documentation on profiling system

## Related to
- #147 (Memory checks before embedding)
- #154 (Dynamic batch sizing)
- #162 (Model loading memory checks)
- This issue provides the foundation for accurate memory management
